File: src/main/java/self/ed/solver/CleverSolver.java
package self.ed.solver;

import static java.util.Collections.emptySet;
import static java.util.Collections.singletonList;
import static java.util.Comparator.comparing;
import static java.util.Optional.ofNullable;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.IntStream.rangeClosed;

import static self.ed.visitor.Visitor.notifyGuessed;
import static self.ed.visitor.Visitor.notifyGuessing;
import static self.ed.visitor.Visitor.notifyInitial;
import static self.ed.visitor.Visitor.notifyOpeningCell;
import static self.ed.visitor.Visitor.notifyOpeningValue;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import self.ed.exception.MultipleSolutionsException;
import self.ed.exception.NoSolutionException;
import self.ed.exception.TimeLimitException;
import self.ed.visitor.Visitor;

public class CleverSolver {

  private final int size;
  private final List<Cell> allCells = new ArrayList<>();
  private final Visitor[] visitors;

  private final List<Cell> pendingCells = new ArrayList<>();
  private final List<Value> pendingValues = new ArrayList<>();

  public CleverSolver(Integer[][] initialValues, Visitor... visitors) {
    this.visitors = visitors;
    size = initialValues.length;
    int blockSize = (int) Math.sqrt(size);
    Map<Integer, Value> valueMap = rangeClosed(1, size)
        .boxed()
        .collect(toMap(Function.identity(), Value::new));

    Map<Cell, Value> openCells = new HashMap<>();
    for (int row = 0; row < size; row++) {
      for (int col = 0; col < size; col++) {
        int block = blockSize * (row / blockSize) + col / blockSize;
        Cell cell = new Cell(row, col, block);
        allCells.add(cell);
        Integer value = initialValues[row][col];
        if (value != null) {
          openCells.put(cell, valueMap.get(value));
        }
      }
    }

    pendingCells.addAll(allCells);
    pendingValues.addAll(valueMap.values());
    pendingCells.forEach(cell -> cell.addCandidates(pendingValues));
    pendingValues.forEach(value -> value.addCandidates(pendingCells));
    notifyInitial(visitors, openCells.size());
    openCells.forEach(Cell::open);
  }

  public Integer[][] solve() {
    if (Thread.interrupted()) {
      throw new TimeLimitException();
    }

//        if (countOpenDistinct(result) < result.length - 1) { //TODO: filter Value.isEmpty() ???
//            throw new MultipleSolutionsException();
//        }

    while (!pendingCells.isEmpty()) {
      try {
        openNext();
      } catch (CannotOpenWithoutGuessingException e) {
        return solveWithGuess(e.getCell(), e.getValue());
      }
    }

    return copyState();
  }

  private void openNext() throws CannotOpenWithoutGuessingException {
    if (pendingValues.isEmpty()) {
      throw new NoSolutionException();
    }

    Cell cell = pendingCells.stream().min(comparing(Cell::countCandidates)).get();
    if (cell.countCandidates() == 1) {
      notifyOpeningCell(visitors);
      cell.open(cell.getCandidate());
      return;
    }

    Value value = pendingValues.stream().min(comparing(Value::countCandidates)).get();
    if (value.countCandidates() == 1) {
      notifyOpeningValue(visitors);
      value.getCandidate().open(value);
      return;
    }

    if (cell.countCandidates() == 0 || value.countCandidates() == 0) {
      throw new NoSolutionException();
    }

    throw new CannotOpenWithoutGuessingException(cell, value);
  }

  private Integer[][] solveWithGuess(Cell cell, Value value) {
    Collection<Cell> guessCells;
    Collection<Value> guessValues;
    if (cell.countCandidates() <= value.countCandidates()) {
      guessCells = singletonList(cell);
      guessValues = cell.getCandidates();
    } else {
      guessCells = value.getCandidates();
      guessValues = singletonList(value);
    }

    List<Integer[][]> solutions = new ArrayList<>();
    notifyGuessing(visitors, guessCells.size() * guessValues.size());
    // System.out.println("Guessing out of " + guessCells.size() * guessValues.size());
    for (Cell guessCell : guessCells) {
      for (Value guessValue : guessValues) {
        Integer[][] nextGuess = copyState();
        nextGuess[guessCell.getRow()][guessCell.getCol()] = guessValue.getValue();
        try {
          solutions.add(new CleverSolver(nextGuess, visitors).solve());
          if (solutions.size() > 1) {
            throw new MultipleSolutionsException();
          }
        } catch (NoSolutionException e) {
          // Our guess did not work, let's try another one
        }
      }
    }

    if (solutions.isEmpty()) {
      throw new NoSolutionException();
    }

    notifyGuessed(visitors);
    return solutions.getFirst();
  }

  private Integer[][] copyState() {
    Integer[][] state = new Integer[size][size];
    allCells.forEach(cell -> state[cell.getRow()][cell.getCol()] = cell.getValue());
    return state;
  }

  @lombok.Value
  @EqualsAndHashCode(callSuper = true)
  private class CannotOpenWithoutGuessingException extends RuntimeException {

    Cell cell;
    Value value;
  }

  @Getter
  @RequiredArgsConstructor
  private class Cell {

    private final int row;
    private final int col;
    private final int block;
    private Value value;
    private Collection<Value> candidates;

    void addCandidates(Collection<Value> candidates) {
      this.candidates = new ArrayList<>(candidates);
    }

    void open(Value value) {
      this.value = value;
      candidates.forEach(candidate -> candidate.removeCandidate(this));
      candidates.clear();
      value.open(this);
      pendingCells.remove(this);
      pendingCells.stream()
          .filter(this::isRelated)
          .forEach(pendingCell -> pendingCell.removeCandidate(value));
    }

    boolean isRelated(Cell cell) {
      return row == cell.row || col == cell.col || block == cell.block;
    }

    void removeCandidate(Value value) {
      value.removeCandidate(this);
      candidates.remove(value);
    }

    int countCandidates() {
      return candidates.size();
    }

    Value getCandidate() {
      return candidates.iterator().next();
    }

    Integer getValue() {
      return ofNullable(value).map(Value::getValue).orElse(null);
    }
  }

  @RequiredArgsConstructor
  private class Value {

    @Getter
    private final Integer value;
    private final Collection<Cell> cells = new ArrayList<>();
    private final Collection<Collection<Cell>> candidates = new ArrayList<>();

    void addCandidates(Collection<Cell> candidates) {
      this.candidates.addAll(candidates.stream().collect(groupingBy(Cell::getRow)).values());
      this.candidates.addAll(candidates.stream().collect(groupingBy(Cell::getCol)).values());
      this.candidates.addAll(candidates.stream().collect(groupingBy(Cell::getBlock)).values());
    }

    void removeCandidate(Cell cell) {
      candidates.forEach(candidateGroup -> candidateGroup.remove(cell));
      candidates.removeIf(Collection::isEmpty);
    }

    void open(Cell cell) {
      removeCandidate(cell);
      cells.add(cell);
      if (cells.size() == size) {
        pendingValues.remove(this);
      }
    }

    int countCandidates() {
      return candidates.stream()
          .mapToInt(Collection::size)
          .min()
          .orElse(0);
    }

    Cell getCandidate() {
      return getCandidates().iterator().next();
    }

    Collection<Cell> getCandidates() {
      return candidates.stream()
          .min(comparing(Collection::size))
          .orElse(emptySet());
    }
  }
}

File: cpp-sudoku/src/clever_solver.cpp
#include "clever_solver.hpp"
#include <cmath>
#include <algorithm>
#include <stdexcept>

namespace sudoku {

CleverSolver::CleverSolver(const std::vector<std::vector<std::optional<int>>>& initialValues) 
    : size(static_cast<int>(initialValues.size())), 
      blockSize(static_cast<int>(std::sqrt(size))) {
    
    // Create value map
    for (int i = 1; i <= size; ++i) {
        valueMap[i] = std::make_unique<Value>(i);
    }
    
    // Create cells and initialize with values
    std::unordered_map<Cell*, Value*> openCells;
    for (int row = 0; row < size; ++row) {
        for (int col = 0; col < size; ++col) {
            int block = blockSize * (row / blockSize) + col / blockSize;
            auto cell = std::make_unique<Cell>(row, col, block);
            allCells.push_back(std::move(cell));
            
            if (initialValues[row][col].has_value()) {
                int value = initialValues[row][col].value();
                openCells[allCells.back().get()] = valueMap[value].get();
            }
        }
    }
    
    // Add candidates to cells and values
    std::vector<Value*> values;
    for (auto& [_, value] : valueMap) {
        values.push_back(value.get());
        pendingValues.push_back(value.get());
    }
    
    std::vector<Cell*> cells;
    for (auto& cell : allCells) {
        cells.push_back(cell.get());
        pendingCells.push_back(cell.get());
    }
    
    for (auto& cell : allCells) {
        cell->addCandidates(values);
    }
    
    for (auto& [_, value] : valueMap) {
        value->addCandidates(cells);
    }
    
    // Open initial cells AFTER setting up all candidates
    for (auto& [cell, value] : openCells) {
        cell->open(value);
        // Remove cell from pending cells
        auto cellToRemove = std::find(pendingCells.begin(), pendingCells.end(), cell);
        if (cellToRemove != pendingCells.end()) {
            pendingCells.erase(cellToRemove);
        }
        // Remove value from related pending cells
        for (auto it = pendingCells.begin(); it != pendingCells.end(); ++it) {
            if (cell->isRelated(*it)) {
                (*it)->removeCandidate(value);
            }
        }
        // Check if value is complete and remove from pending values
        if (value->cells.size() == static_cast<size_t>(size)) {
            auto valueToRemove = std::find(pendingValues.begin(), pendingValues.end(), value);
            if (valueToRemove != pendingValues.end()) {
                pendingValues.erase(valueToRemove);
            }
        }
    }
}

std::vector<std::vector<std::optional<int>>> CleverSolver::solve() {
    while (!pendingCells.empty()) {
        try {
            openNext();
        } catch (const CannotOpenWithoutGuessingException& e) {
            return solveWithGuess(e.cell, e.value);
        }
    }
    return copyState();
}

void CleverSolver::openNext() {
    if (pendingValues.empty()) {
        throw NoSolutionException();
    }
    
    // Find cell with minimum candidates
    auto cellIt = std::min_element(pendingCells.begin(), pendingCells.end(),
        [](const Cell* a, const Cell* b) {
            return a->countCandidates() < b->countCandidates();
        });
    
    Cell* cell = *cellIt;
    
    if (cell->countCandidates() == 1) {
        Value* valueToOpen = cell->getCandidate();
        cell->open(valueToOpen);
        // Remove cell from pending cells
        pendingCells.erase(cellIt);
        // Remove value from related pending cells
        for (auto it = pendingCells.begin(); it != pendingCells.end(); ++it) {
            if (cell->isRelated(*it)) {
                (*it)->removeCandidate(valueToOpen);
            }
        }
        // Check if value is complete and remove from pending values
        if (valueToOpen->cells.size() == static_cast<size_t>(size)) {
            auto valueToRemove = std::find(pendingValues.begin(), pendingValues.end(), valueToOpen);
            if (valueToRemove != pendingValues.end()) {
                pendingValues.erase(valueToRemove);
            }
        }
        return;
    }
    
    // Find value with minimum candidates
    auto valueIt = std::min_element(pendingValues.begin(), pendingValues.end(),
        [](const Value* a, const Value* b) {
            return a->countCandidates() < b->countCandidates();
        });
    
    Value* value = *valueIt;
    
    if (value->countCandidates() == 1) {
        Cell* targetCell = value->getCandidate();
        targetCell->open(value);
        // Remove cell from pending cells
        auto cellToRemove = std::find(pendingCells.begin(), pendingCells.end(), targetCell);
        if (cellToRemove != pendingCells.end()) {
            pendingCells.erase(cellToRemove);
        }
        // Remove value from related pending cells
        for (auto it = pendingCells.begin(); it != pendingCells.end(); ++it) {
            if (targetCell->isRelated(*it)) {
                (*it)->removeCandidate(value);
            }
        }
        // Check if value is complete and remove from pending values
        if (value->cells.size() == static_cast<size_t>(size)) {
            auto valueToRemove = std::find(pendingValues.begin(), pendingValues.end(), value);
            if (valueToRemove != pendingValues.end()) {
                pendingValues.erase(valueToRemove);
            }
        }
        return;
    }
    
    if (cell->countCandidates() == 0 || value->countCandidates() == 0) {
        throw NoSolutionException();
    }
    
    throw CannotOpenWithoutGuessingException(cell, value);
}

std::vector<std::vector<std::optional<int>>> CleverSolver::solveWithGuess(Cell* cell, Value* value) {
    std::vector<Cell*> guessCells;
    std::vector<Value*> guessValues;
    
    if (cell->countCandidates() <= value->countCandidates()) {
        guessCells = {cell};
        guessValues.assign(cell->candidates.begin(), cell->candidates.end());
    } else {
        auto candidates = value->getCandidates();
        guessCells.assign(candidates.begin(), candidates.end());
        guessValues = {value};
    }
    
    std::vector<std::vector<std::vector<std::optional<int>>>> solutions;
    
    for (Cell* guessCell : guessCells) {
        for (Value* guessValue : guessValues) {
            auto nextGuess = copyState();
            nextGuess[guessCell->row][guessCell->col] = guessValue->value;
            
            try {
                solutions.push_back(CleverSolver(nextGuess).solve());
                if (solutions.size() > 1) {
                    throw MultipleSolutionsException();
                }
            } catch (const NoSolutionException&) {
                // Our guess did not work, try another one
            }
        }
    }
    
    if (solutions.empty()) {
        throw NoSolutionException();
    }
    
    return solutions[0];
}

std::vector<std::vector<std::optional<int>>> CleverSolver::copyState() const {
    std::vector<std::vector<std::optional<int>>> state(size, std::vector<std::optional<int>>(size));
    for (const auto& cell : allCells) {
        state[cell->row][cell->col] = cell->getValue();
    }
    return state;
}

// Cell methods
void CleverSolver::Cell::open(Value* val) {
    this->value = val;
    for (Value* candidate : candidates) {
        candidate->removeCandidate(this);
    }
    candidates.clear();
    val->open(this);
    
    // Remove this cell from pending cells
    // Note: This will be handled by the solver when it calls openNext()
    
    // Remove this value from all related pending cells
    // This is the key part that was missing!
}

bool CleverSolver::Cell::isRelated(const Cell* cell) const {
    return row == cell->row || col == cell->col || block == cell->block;
}

void CleverSolver::Cell::removeCandidate(Value* val) {
    val->removeCandidate(this);
    candidates.erase(val);
}

// Value methods
void CleverSolver::Value::addCandidates(const std::vector<Cell*>& cells) {
    // Group by row
    std::unordered_map<int, std::unordered_set<Cell*>> rowGroups, colGroups, blockGroups;
    
    for (Cell* cell : cells) {
        rowGroups[cell->row].insert(cell);
        colGroups[cell->col].insert(cell);
        blockGroups[cell->block].insert(cell);
    }
    
    for (auto& [_, group] : rowGroups) {
        candidates.push_back(group);
    }
    for (auto& [_, group] : colGroups) {
        candidates.push_back(group);
    }
    for (auto& [_, group] : blockGroups) {
        candidates.push_back(group);
    }
}

void CleverSolver::Value::removeCandidate(Cell* cell) {
    for (auto& candidateGroup : candidates) {
        candidateGroup.erase(cell);
    }
    candidates.erase(
        std::remove_if(candidates.begin(), candidates.end(),
            [](const std::unordered_set<Cell*>& group) { return group.empty(); }),
        candidates.end()
    );
}

void CleverSolver::Value::open(Cell* cell) {
    removeCandidate(cell);
    cells.insert(cell);
}

size_t CleverSolver::Value::countCandidates() const {
    if (candidates.empty()) return 0;
    return std::min_element(candidates.begin(), candidates.end(),
        [](const std::unordered_set<Cell*>& a, const std::unordered_set<Cell*>& b) {
            return a.size() < b.size();
        })->size();
}

CleverSolver::Cell* CleverSolver::Value::getCandidate() const {
    auto minGroup = std::min_element(candidates.begin(), candidates.end(),
        [](const std::unordered_set<Cell*>& a, const std::unordered_set<Cell*>& b) {
            return a.size() < b.size();
        });
    return *minGroup->begin();
}

std::unordered_set<CleverSolver::Cell*> CleverSolver::Value::getCandidates() const {
    if (candidates.empty()) return {};
    auto minGroup = std::min_element(candidates.begin(), candidates.end(),
        [](const std::unordered_set<Cell*>& a, const std::unordered_set<Cell*>& b) {
            return a.size() < b.size();
        });
    return *minGroup;
}

} // namespace sudoku
